### PR TITLE
examples/seq2seq supports translation

### PR DESCRIPTION
**Pull request by sshleifer**
- renames `examples/summarization` -> `examples/seq2seq`
- finetune.py and run_eval.py support mbart, marian and t5.
- task_specific_params are used
- if you specify task='translation', then your metric becomes BLEU instead of ROUGE.
- improved `README.md`
- lots of test coverage
- scripts to reproduce distilbart results

TODO:
- [x] verified distilbart commands replicate posted results.
- [x] new xsum shared task URL.
- [x] mini models for marian
- [x] mbart finetuning unittests.

Postponed and made issues for:
- [ ] check bleu scores for translation models with run_eval.py

### CODE REVIEW

File: examples/seq2seq/finetune.py [hunk 1]
```diff
@@ -89,12 +93,12 @@ def __init__(self, hparams, **kwargs):
 
     def freeze_embeds(self):
         """Freeze token embeddings and positional embeddings for bart, just token embeddings for t5."""
-        if self.model.config.model_type == "bart":
+        try:
```
Comment (sshleifer) [inline examples/seq2seq/finetune.py:96]: this should recursively search for `nn.Embedding`

File: examples/seq2seq/finetune.py [hunk 2]
```diff
@@ -130,20 +134,24 @@ def validation_epoch_end(self, outputs, prefix="val") -> Dict:
         self.step_count += 1
         losses = {k: torch.stack([x[k] for x in outputs]).mean() for k in self.loss_names}
         loss = losses["loss"]
-        rouges = {k: np.array([x[k] for x in outputs]).mean() for k in ROUGE_KEYS + ["gen_time", "summ_len"]}
-        rouge_tensor: torch.FloatTensor = torch.tensor(rouges["rouge2"]).type_as(loss)
+        rouges = {k: np.array([x[k] for x in outputs]).mean() for k in self.metric_names + ["gen_time", "summ_len"]}
```
Comment (sshleifer) [inline examples/seq2seq/finetune.py:137]: bad var name

File: examples/seq2seq/run_eval.py [hunk 3]
```diff
@@ -70,8 +72,7 @@ def run_generate():
         output_lns = [x.rstrip() for x in open(args.output_path).readlines()]
         reference_lns = [x.rstrip() for x in open(args.reference_path).readlines()]
 
-        rouge: dict = calculate_rouge(output_lns, reference_lns)
-
+        rouge: dict = score_fn(output_lns, reference_lns)
```
Comment (sshleifer) [inline examples/seq2seq/run_eval.py:75]: bad varname

File: examples/seq2seq/test_summarization_examples.py [hunk 4]
```diff
@@ -114,6 +115,20 @@ def test_bdc_multigpu(self):
         )
         self._bart_distiller_cli(updates)
 
+    @unittest.skip(MSG)
+    def test_bdc_t5_eval_fp16(self):
+        updates = dict(
+            fp16=FP16_EVER,
```
Comment (sshleifer) [inline examples/seq2seq/test_summarization_examples.py:121]: constant should be called 'CUDA_AVAILABLE'

File: examples/seq2seq/test_summarization_examples.py [hunk 5]
```diff
@@ -114,6 +115,20 @@ def test_bdc_multigpu(self):
...
+            gpus=1,
+            model_type="t5",
```
Comment (sshleifer) [inline examples/seq2seq/test_summarization_examples.py:123]: del model_type, its unused

File: examples/seq2seq/utils.py [hunk 6]
```diff
@@ -95,6 +98,8 @@ def __init__(
             tok_name=tok_name,
         )
         tgt_path = os.path.join(data_dir, type_path + ".target")
+        if hasattr(tokenizer, "set_lang"):
+            tokenizer.set_lang("ro_RO")
```
Comment (sshleifer) [inline examples/seq2seq/utils.py:102]: delete this!

File: src/transformers/tokenization_bart.py [hunk 7]
```diff
@@ -115,6 +115,10 @@ def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None) -> Lis
         # We don't expect to process pairs, but leave the pair logic for API consistency
         return token_ids_0 + token_ids_1 + special_tokens
 
+    def set_lang(self, lang: str) -> None:
```
Comment (sshleifer) [inline src/transformers/tokenization_bart.py:118]: hack to allow setting language code and then calling batch_encode_plus for target language.

**Comment by sshleifer** [comment]
Gunna merge this to avoid tweeting broken links. 
Comments welcome, I will probably need to do more cleanup.
